src/metabase/handler.clj

```diff
@@ -23,34 +24,45 @@
-  (->
-   ;; in production, dereference routes now because they will not change at runtime, so we don't need to waste time
-   ;; dereferencing the var on every request. For dev & test, use the var instead so it can be tweaked without having
-   ;; to restart the web server
-   (if config/is-prod?
-     routes/routes
-     #'routes/routes)
-   ;; ▼▼▼ POST-PROCESSING ▼▼▼ happens from TOP-TO-BOTTOM
-   mw.exceptions/catch-uncaught-exceptions ; catch any Exceptions that weren't passed to `raise`
-   mw.exceptions/catch-api-exceptions      ; catch exceptions and return them in our expected format
-   mw.log/log-api-call
-   mw.security/add-security-headers        ; Add HTTP headers to API responses to prevent them from being cached
-   mw.json/wrap-json-body                  ; extracts json POST body and makes it avaliable on request
-   mw.json/wrap-streamed-json-response     ; middleware to automatically serialize suitable objects as JSON in responses
-   wrap-keyword-params                     ; converts string keys in :params to keyword keys
-   wrap-params                             ; parses GET and POST params as :query-params/:form-params and both as :params
-   mw.misc/maybe-set-site-url              ; set the value of `site-url` if it hasn't been set yet
-   mw.session/bind-current-user            ; Binds *current-user* and *current-user-id* if :metabase-user-id is non-nil
-   mw.session/wrap-current-user-info       ; looks for :metabase-session-id and sets :metabase-user-id and other info if Session ID is valid
-   mw.session/wrap-session-id              ; looks for a Metabase Session ID and assoc as :metabase-session-id
-   mw.auth/wrap-api-key                    ; looks for a Metabase API Key on the request and assocs as :metabase-api-key
-   wrap-cookies                            ; Parses cookies in the request map and assocs as :cookies
-   mw.misc/add-content-type                ; Adds a Content-Type header for any response that doesn't already have one
-   mw.misc/disable-streaming-buffering     ; Add header to streaming (async) responses so ngnix doesn't buffer keepalive bytes
-   wrap-gzip                               ; GZIP response if client can handle it
-   mw.misc/bind-request                    ; bind `metabase.middleware.misc/*request*` for the duration of the request
-   mw.ssl/redirect-to-https-middleware))   ; Redirect to HTTPS if configured to do so
-;; ▲▲▲ PRE-PROCESSING ▲▲▲ happens from BOTTOM-TO-TOP
+  (apply-middleware routes/routes))
+
+;; during interactive dev, recreate `app` whenever a middleware var or `routes/routes` changes.
+(when config/is-dev?
+  (doseq [varr  (cons #'routes/routes middleware)
+          :when (instance? clojure.lang.IRef varr)]
+    (add-watch varr ::reload (fn [_ _ _ _]
+                               (printf "%s changed, rebuilding %s" varr #'app)
+                               (alter-var-root #'app (constantly (apply-middleware routes/routes)))))))
```

Comment:
note to self: we should merge this improvement regardless of whether we do anything with the rest of this PR